commit 96a7109a16665255b65d021e24141c2edae0e202
Author: Johan Almbladh <johan.almbladh@anyfinetworks.com>
Date:   Thu Apr 1 18:44:55 2021 +0200

    mac80211: Set priority and queue mapping for injected frames
    
    Some drivers, for example mt76, use the skb priority field, and
    expects that to be consistent with the skb queue mapping. On some
    frame injection code paths that was not true, and it broke frame
    injection. Now the skb queue mapping is set according to the skb
    priority value when the frame is injected. The skb priority value
    is also derived from the frame data for all frame types, as it
    was done prior to commit dbd50a851c50 (only allocate one queue
    when using iTXQs). Fixes frame injection with the mt76 driver on
    MT7610E chipset.
    
    Signed-off-by: Johan Almbladh <johan.almbladh@anyfinetworks.com>
    Link: https://lore.kernel.org/r/20210401164455.978245-1-johan.almbladh@anyfinetworks.com
    Signed-off-by: Johannes Berg <johannes.berg@intel.com>

diff --git a/net/mac80211/tx.c b/net/mac80211/tx.c
index 7aaa71ed49b0..3d6c5b8ec0c2 100644
--- a/net/mac80211/tx.c
+++ b/net/mac80211/tx.c
@@ -2276,17 +2276,6 @@ netdev_tx_t ieee80211_monitor_start_xmit(struct sk_buff *skb,
 						    payload[7]);
 	}
 
-	/* Initialize skb->priority for QoS frames. If the DONT_REORDER flag
-	 * is set, stick to the default value for skb->priority to assure
-	 * frames injected with this flag are not reordered relative to each
-	 * other.
-	 */
-	if (ieee80211_is_data_qos(hdr->frame_control) &&
-	    !(info->control.flags & IEEE80211_TX_CTRL_DONT_REORDER)) {
-		u8 *p = ieee80211_get_qos_ctl(hdr);
-		skb->priority = *p & IEEE80211_QOS_CTL_TAG1D_MASK;
-	}
-
 	rcu_read_lock();
 
 	/*
@@ -2350,6 +2339,15 @@ netdev_tx_t ieee80211_monitor_start_xmit(struct sk_buff *skb,
 
 	info->band = chandef->chan->band;
 
+	/* Initialize skb->priority according to frame type and TID class,
+	 * with respect to the sub interface that the frame will actually
+	 * be transmitted on. If the DONT_REORDER flag is set, the original
+	 * skb-priority is preserved to assure frames injected with this
+	 * flag are not reordered relative to each other.
+	 */
+	ieee80211_select_queue_80211(sdata, skb, hdr);
+	skb_set_queue_mapping(skb, ieee80211_ac_from_tid(skb->priority));
+
 	/* remove the injection radiotap header */
 	skb_pull(skb, len_rthdr);